application/controllers/OptinController.php

```diff
@@ -27,53 +27,183 @@ class OptinController extends LSYii_Controller
+
+        if (!$surveyId) {
+            $this->redirect(['/']);
+        }
+        $surveyId = (int) $surveyId; //Make sure it's an integer (protect from SQL injects)
+        $survey = Survey::model()->findByPk($surveyId);
 
         //Check that there is a SID
         // Get passed language from form, so that we dont loose this!
-        if (!isset($sLanguageCode) || $sLanguageCode == "" || !$sLanguageCode) {
-            $sBaseLanguage = $oSurvey->language;
+        if (!isset($languageCode) || $languageCode == "" || !$languageCode) {
+            $baseLanguage = $survey->language;
         } else {
-            $sBaseLanguage = sanitize_languagecode($sLanguageCode);
+            $baseLanguage = sanitize_languagecode($languageCode);
         }
 
-        Yii::app()->setLanguage($sBaseLanguage);
+        Yii::app()->setLanguage($baseLanguage);
 
-        if (empty($oSurvey) || !$oSurvey->hasTokensTable) {
+        if (empty($survey) || !$survey->hasTokensTable) {
             throw new CHttpException(404, "This survey does not seem to exist. It may have been deleted or the link you were given is outdated or incorrect.");
         } else {
-            LimeExpressionManager::singleton()->loadTokenInformation($iSurveyID, $sToken, false);
-            $oToken = Token::model($iSurveyID)->findByAttributes(array('token' => $sToken));
+            LimeExpressionManager::singleton()->loadTokenInformation($surveyId, $accessToken, false);
+            $token = Token::model($surveyId)->findByAttributes(['token' => $accessToken]);
 
-            if (!isset($oToken)) {
-                $sMessage = gT('You are not a participant of this survey.');
+            if (!isset($token)) {
+                $message = gT('You are not a participant of this survey.');
             } else {
-                if ($oToken->emailstatus == 'OptOut') {
-                    $oToken->emailstatus = 'OK';
-                    $oToken->save();
-                    $sMessage = gT('You have been successfully added back to this survey.');
-                } elseif ($oToken->emailstatus == 'OK') {
-                    $sMessage = gT('You are already a participant of this survey.');
+                if ($token->emailstatus == 'OptOut') {
+                    $token->emailstatus = 'OK';
+                    $token->save();
+                    $message = gT('You have been successfully added back to this survey.');
+                } elseif ($token->emailstatus == 'OK') {
+                    $message = gT('You are already a participant of this survey.');
                 } else {
-                    $sMessage = gT('You have been already removed from this survey.');
+                    $message = gT('You have been already removed from this survey.');
+                }
+                // If the $global param is true and 'allowunblacklist' is enabled, remove from the blacklist
+                if ($global && Yii::app()->getConfig('allowunblacklist') == "Y") {
```

Comment:
Yii::app()->getConfig('allowunblacklist') == "Y" ?

We use true/false or 1/0 for all other boolean settings …

Please : use boolean

Reply:
This was carried over code... but I get your point.
This is saved on the DB as Y/N.
Changing it could be very painfull.

Reply:
Since when its saved as Y/N ? 

In 3.X : it was 0/1
![Capture d’écran du 2022-02-08 16-18-25](https://user-images.githubusercontent.com/1439428/153017249-bcfa53e1-7fa4-4140-95a5-cf4726333fd8.png)

And it's clearly better

Reply:
Else : it can be saved as Y for true, since it was saved as 0 for false : we can use boolval :)

Reply:
Argls … show as 0, saved a N ...

**awfull** 😵

I close